Make clippy a subrepo (#70650)
rust-lang/rust: CONTRIBUTING.md

Speaking of tests, Rust has a comprehensive test suite. More information about
it can be found [here][rctd].

### External Dependencies (subrepo)

As a developer to this repository, you don't have to treat the following external projects
differently from other crates that are directly in this repo:

* [clippy](https://github.com/rust-lang/rust-clippy)

> They are just regular files and directories. This is in contrast to `submodule` dependencies.

Reviewer: ```suggestion
They are just regular files and directories. This is in contrast to `submodule` dependencies
(see below for those).
```

If you want to synchronize or otherwise work with subrepos, install the `git subrepo` command via
instructions found at https://github.com/ingydotnet/git-subrepo

> #### Creating a new subrepo dependency

Reviewer: Contributors should rarely have to do this, correct? If so, then it should be mentioned last, and it should be said that this is only for *adding new tools to rustc*.

If you want to create a new subrepo dependency from an existing repository, call (from this
repository's root directory!!)

```
git subrepo clone https://github.com/rust-lang/rust-clippy.git src/tools/clippy
```

This will create a new commit, which you may not rebase under any circumstances! Delete the commit
and redo the operation if you need to rebase.

Now you're done, the `src/tools/clippy` directory behaves as if clippy were part of the rustc
monorepo, so no one but you (or others that synchronize subrepos) needs to have `git subrepo`
installed.

> #### Synchronizing a subrepo

Reviewer: This, too, is something that contributors *usually* shouldn't have to do, instead that would generally be tool maintainer's job?

Is there a way for us maintainers to be notified when our tools change in rustc so that we know we need to sync? (Longer-term, having a bot to also do the syncing would be even better of course. But some kind of notification is required, I think, to avoid silent divergence.)

Author: Contributors shouldn't care about this entire section. Where else do I put it?

We could probably extend the toolstate bot to check whether `git diff src/tools/clippy` has any output, and in that case ping the team

> There are two synchronization directions: push and pull.

Reviewer: I initially was very confused by the overlap of terminology between `git pull` and `git subrepo pull`.

So maybe it is better to say "subrepo push/pull" each and every time that kind of push/pull is meant?

A `git subrepo push src/tools/clippy`
takes all the changes that
happened to the copy in this repo and creates commits on the remote repo that match the local

> changes (so ever local commit that touched the subrepo causes a commit on the remote repo). Again,

Reviewer: ```suggestion
changes (so every local commit that touched the subrepo causes a commit on the remote repo). Again,
```

> Even this `push` operation creates a commit in this repo that you need to get merged without

Reviewer: ```suggestion
this `push` operation also creates a commit in this (rustc) repo that you need to get merged without
```

> rebasing. This is very important in order to make future synchronizations work.

Reviewer: So the idea is, I will `git subrepo push`, and then make a rustc PR (likely self-r+ p=1) and a tool PR? And then what, I wait until the rustc PR lands before landing the tool PR, or vice versa? Either way there is a chance of conflicts where one PR lands and the other one does not, and it is unclear how bad that is. We should give some general guidance to at least have a consistent process.

Author: We have no experience here. Not sure how to build guidance without really knowing how it will behave in practice.

Reviewer: Come up with a best guess, adjust it later if it doesn't work? ;)

A `git subrepo pull src/tools/clippy` takes all changes since the last `pull` from the clippy
repo and creates a single commit in the rustc repo with all the changes. Again, do not rebase this

> commit under any circumstances. Redo the operation if you need to.

Reviewer: Is there a way to not have to re-solve all conflicts in case of a rebase? Would be quite annoying otherwise.

Author: You can still merge, just not rebase. Just don't rebase ;)

Reviewer: The text advises me to redo if I have to rebase. Maybe it should advise me to merge then?

I suppose a conflict can only arise if there is another subrepo pull/push for the same repo, which makes sense -- syncs have to be sequentialized.

#### Throwing the towel (aka, everything is broken)

> If the subrepo has ended up in an unusable state out of any reason

Reviewer: This looks incomplete.
